Remove XCPretty dependency and add xcbeautify as an option

joshdholtz opened the pull request:
## Project
👉 https://github.com/fastlane/fastlane/projects/4

## How To Test

Install `xcbeautify` by `brew install xcbeautify` or other methods listed [here](https://github.com/tuist/xcbeautify)

Test the release candidate at `2.201.0.rc1`

```sh
gem install fastlane -v "2.201.0.rc1"
```

```rb
gem "fastlane", "2.201.0.rc1"
```

ℹ️  **Please give any success stories, feedback, or issues in the comments of this PR!**

## Goal
- Remove [XCPretty](https://github.com/xcpretty/xcpretty) as a **needed** dependency
- Keep it as an option for `xcodebuild` formatting
- Open up support for other formatters like [xcbeautify](https://github.com/thii/xcbeautify)

## Changes

### `rc1`
- `scan` no longer relies on [xcpretty](https://github.com/xcpretty/xcpretty) for anything
  - [trainer](https://github.com/fastlane-community/trainer) has now been moved inside of _fastlane_ and is used to determine determine success and failure of `scan`
- `scan` option for `number_of_retries` now uses Xcode 13's retry tests 
  - In Xcode 12 and under it will still use _fastlane_'s implementation to try to only rerun tests that have failed
- Added `xcodebuild_formatter` option to `scan`, `gym`, and `snapshot`
  - Defaults to [xcbeautify](https://github.com/tuist/xcbeautify)
    - But will default back to `xcpretty` if any "xcpretty" specific options are set (ex: `xcpretty_formatter`, `xcpretty_args`)
  - Otherwise falls back to [xcpretty](https://github.com/xcpretty/xcpretty)
  - You can also pass anything for `xcodebuild` to pipe to like `xcodebuild_formatter: "/path/to/other/xcbeautify"`

### `rc2`
- Added doc to `xcodebuild_formatter` description that says no formatter will be used if set with empty string
- `output_types` are fixed and now being generated
  - `junit` will be made by `xcpretty` is using `xcpretty`, otherwise it will be made with `trainer`
  - `html` will only be made if using `xcpretty`

### `rc3`
- Removed invalid `output_types` error and replaced with warnings and how to fix
  - Adjusting default (based on `xcodebuild_formatter`) felt wrong and could easily cause worse issues
-  Prevent `trainer` from running if `build_for_testing: true` is set
- Fixed junit/xml output so that it doesn't show retries in the total counts
  - Ex: failures will show as `0` if there were retries that eventually passed
- Fix for `trainer`'s plist parsing to return same data for retries as the xcresult parsing 

### Release
- Added new option to `scan` for `output_remove_retry_attempts`
  - Removes the count and any retry attempts from junit report and scan test results table
- Added migration error if using `fastlane-plugin-trainer` and tells user to remove it

## Examples

```rb
scan(
  xcodebuild_formatter: "xcbeautify"
)
```

## Further Work
There might be some things like `scan`'s `output_types` that may still need some work 😇 

## Pull Requests
- https://github.com/fastlane/fastlane/pull/19636
- https://github.com/fastlane/fastlane/pull/19695
- https://github.com/fastlane/fastlane/pull/19774
- https://github.com/fastlane/fastlane/pull/19778
- https://github.com/fastlane/fastlane/pull/19786

joshdholtz committed:
Bring in trainer as an action and standalone tool because trainer logic is needed for scan (#19624)

* Bring in trainer as an action and standalone tool because trainer logic is needed for scan

* Add action category

* This should fix failing tests

* UGH THIS

* Maybe this time

joshdholtz committed:
Bring in trainer as an action and standalone tool because trainer logic is needed for scan (#19624)

* Bring in trainer as an action and standalone tool because trainer logic is needed for scan

* Add action category

* This should fix failing tests

* UGH THIS

* Maybe this time

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Nowlin2 commented:
SwifterSwift is a collection of over 500 native Swift extensions, with handy methods, syntactic sugar, and performance improvements for wide range of primitive data types.

joshdholtz committed:
Replace scan success factor from xcpretty to trainer (#19636)

* Start of thing

* Tests pass again woot woot

* Remove todos

* Fix other failing tests

* omg tests are hard

* Some small fixes from PR review

* Improvements for table output

* Fix trainer tests

joshdholtz committed:
Replace scan success factor from xcpretty to trainer (#19636)

* Start of thing

* Tests pass again woot woot

* Remove todos

* Fix other failing tests

* omg tests are hard

* Some small fixes from PR review

* Improvements for table output

* Fix trainer tests

joshdholtz committed:
Add native xcode 13 retries (#19695)

* Add native xcode 13 retries

* Clean if statement

joshdholtz committed:
Add native xcode 13 retries (#19695)

* Add native xcode 13 retries

* Clean if statement

joshdholtz committed:
Allow scan to use xcbeautify

joshdholtz committed:
Added tests

joshdholtz committed:
Fix description of xcodebuild_formatter

joshdholtz committed:
Start of allowing xcbeautify in gym

joshdholtz committed:
Fix tests for ubuntu

joshdholtz committed:
[xcbeautify][scan] allow scan to use xcbeautify (#19774)

* Allow scan to use xcbeautify

* Added tests

* Fix description of xcodebuild_formatter

* Fix tests for ubuntu

joshdholtz committed:
This is good now

joshdholtz committed:
Merge

joshdholtz committed:
Fix test for ubuntu

joshdholtz committed:
Actually fix ubuntu tests

joshdholtz committed:
Somehow I really messed up merge of scan and gym xbeautify integration

joshdholtz committed:
[xcbeautify][snapshot] snapshot allow xcbeautify (#19786)

joshdholtz committed:
Merge remote-tracking branch 'origin/master' into remove-xcpretty-dependency-add-xcbeautify-option

joshdholtz committed:
2.201.0.rc1

sebskuse commented:
First up, thanks for adding this! We've been using a workaround to enable `xcbeautify` with `fastlane` (as we use parallel UI testing extensively) for a while.

I've removed our workarounds and switched over to the `rc1` tag. I'm seeing this correctly detect success and failure states and show output correctly using `xcbeautify` as with our previous workaround. The only piece of odd behaviour I've seen so far is at the end of `scan` it prints:

`Tests Passed: 0 failed, 9 total (0.669 seconds)`

... where in reality it's executed a few thousand tests and taken (sadly) a lot longer than that.

chojnac commented on scan/lib/scan/runner.rb:
```diff
@@ -135,27 +137,68 @@ def retryable_tests(input)
+        path: result_bundle_path,
+        output_path: output_path,
+        silent: true,
+        extension: "xml"
```
this should be `.xml` otherwise the result file is `schemaxml`

chojnac commented:
Thanks for the fantastic work on this PR!

In my tests I've noticed a few issues:
- when `scan` run with `test_without_building: true` option command fails with `fastlane-2.201.0.rc1/scan/lib/scan/runner.rb:163:in `block in trainer_test_results': \e[31m[!] undefined method `[]' for true:TrueClass\e[0m (NoMethodError)` message. Looks like trainer shouldn't be run if there are no test runned. 
- trainer junit output filename is missing a dot - got `SomeSchemaxml` should be `SomeSchema.xml`

AvdLee commented on gym/lib/gym/options.rb:
```diff
@@ -230,8 +231,18 @@ def self.plain_options
+
+        FastlaneCore::ConfigItem.new(key: :xcodebuild_formatter,
+                                     env_names: ["GYM_XCODEBUILD_FORMATTER", "FASTLANE_XCODEBUILD_FORMATTER"],
+                                     description: "xcodebuild formatter to use (ex: 'xcbeautify', 'xcbeautify --quieter', 'xcpretty', 'xcpretty -test')",
```
Can we also use no formatter? `disable_xcpretty` is now deprecated in favour of `xcodebuild_formatter`, but it does not seem to come with a disabled option?

joshdholtz replied:
@AvdLee Oops! That description is supposed to mention just set to an empty string to disable output 🤦‍♂️ I'll update that and make sure that works for `gym` and the other tools!

Kylmakalle commented:
Experiencing same issue
> * when `scan` run with `test_without_building: true` option command fails with `fastlane-2.201.0.rc1/scan/lib/scan/runner.rb:163:in `block in trainer_test_results': \e[31m[!] undefined method `[]' for true:TrueClass\e[0m (NoMethodError)` message. Looks like trainer shouldn't be run if there are no test runned.

joshdholtz commented:
> First up, thanks for adding this! We've been using a workaround to enable `xcbeautify` with `fastlane` (as we use parallel UI testing extensively) for a while.
> 
> I've removed our workarounds and switched over to the `rc1` tag. I'm seeing this correctly detect success and failure states and show output correctly using `xcbeautify` as with our previous workaround. The only piece of odd behaviour I've seen so far is at the end of `scan` it prints:
> 
> `Tests Passed: 0 failed, 9 total (0.669 seconds)`
> 
> ... where in reality it's executed a few thousand tests and taken (sadly) a lot longer than that.

Weird! I wonder if its picking up a different `xcresult` somewhere 🤔  Are you able to send me all the options you are setting on `scan` and possibly your console output if you can?

joshdholtz commented:
@chojnac @Kylmakalle I'll get a fix out for `test_without_build: true` into `rc2` for you 💪  I'll ping you when its released!

sebskuse commented:
> > First up, thanks for adding this! We've been using a workaround to enable `xcbeautify` with `fastlane` (as we use parallel UI testing extensively) for a while.
> > I've removed our workarounds and switched over to the `rc1` tag. I'm seeing this correctly detect success and failure states and show output correctly using `xcbeautify` as with our previous workaround. The only piece of odd behaviour I've seen so far is at the end of `scan` it prints:
> > `Tests Passed: 0 failed, 9 total (0.669 seconds)`
> > ... where in reality it's executed a few thousand tests and taken (sadly) a lot longer than that.
> 
> @sebskuse Weird! I wonder if its picking up a different `xcresult` somewhere 🤔 Are you able to send me all the options you are setting on `scan` and possibly your console output if you can?

This one was on one of our ephemeral CI machines, so no possibility of anything else being picked up. Can provide you with some more info though! Have attached the scan table below. Another thing I've noticed is because I've now removed output_types (previously we had this set to `output_types('')` as a workaround - presumed we no longer need this?) it's now being set to `junit,html`. Not sure if that matters for anything.

<details>
   <summary>scan output</summary>

```
+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                                                            Summary for scan 2.201.0.rc1                                                                                                            |
+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| scheme                                         | OurScheme                                                                                                                                                                                         |
| buildlog_path                                  | fastlane/output/                                                                                                                                                                                  |
| output_directory                               | fastlane/test_output/OurScheme                                                                                                                                                                    |
| testplan                                       | Integration                                                                                                                                                                                       |
| skip_build                                     | true                                                                                                                                                                                              |
| workspace                                      | OurWorkspace.xcworkspace                                                                                                                                                                          |
| device                                         | iPhone 11 Pro (15.2)                                                                                                                                                                              |
| skip_detect_devices                            | false                                                                                                                                                                                             |
| ensure_devices_found                           | false                                                                                                                                                                                             |
| force_quit_simulator                           | false                                                                                                                                                                                             |
| reset_simulator                                | false                                                                                                                                                                                             |
| disable_slide_to_type                          | true                                                                                                                                                                                              |
| reinstall_app                                  | false                                                                                                                                                                                             |
| app_identifier                                 | uk.co.us.ourApp                                                                                                                                                                                   |
| clean                                          | false                                                                                                                                                                                             |
| code_coverage                                  | true                                                                                                                                                                                              |
| open_report                                    | false                                                                                                                                                                                             |
| output_types                                   | html,junit                                                                                                                                                                                        |
| include_simulator_logs                         | false                                                                                                                                                                                             |
| xcodebuild_formatter                           | xcbeautify                                                                                                                                                                                        |
| derived_data_path                              | build                                                                                                                                                                                             |
| should_zip_build_products                      | false                                                                                                                                                                                             |
| output_xctestrun                               | false                                                                                                                                                                                             |
| result_bundle                                  | true                                                                                                                                                                                              |
| use_clang_report_name                          | false                                                                                                                                                                                             |
| disable_concurrent_testing                     | false                                                                                                                                                                                             |
| xcargs                                         | GCC_TREAT_WARNINGS_AS_ERRORS=YES SWIFT_TREAT_WARNINGS_AS_ERRORS=YES OTHER_SWIFT_FLAGS='$(inherited) -Xfrontend -warn-long-function-bodies=700 -Xfrontend -warn-long-expression-type-checking=700' |
| app_name                                       | OurApp                                                                                                                                                                                            |
| slack_use_webhook_configured_username_and_icon | false                                                                                                                                                                                             |
| slack_username                                 | fastlane                                                                                                                                                                                          |
| slack_icon_url                                 | https://fastlane.tools/assets/img/fastlane_icon.png                                                                                                                                               |
| skip_slack                                     | true                                                                                                                                                                                              |
| slack_only_on_failure                          | false                                                                                                                                                                                             |
| xcodebuild_command                             | env NSUnbufferedIO=YES xcodebuild                                                                                                                                                                 |
| skip_package_dependencies_resolution           | true                                                                                                                                                                                              |
| disable_package_automatic_updates              | false                                                                                                                                                                                             |
| use_system_scm                                 | false                                                                                                                                                                                             |
| number_of_retries                              | 0                                                                                                                                                                                                 |
| fail_build                                     | false                                                                                                                                                                                             |
| xcode_path                                     | /Applications/Xcode.app                                                                                                                                                                           |
+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
</details>

which results in the following xcodebuild command:

<details>
   <summary>xcodebuild</summary>

```
set -o pipefail && env NSUnbufferedIO=YES xcodebuild -workspace OurWorkspace.xcworkspace -scheme OurScheme -derivedDataPath build -destination 'platform=iOS Simulator,id=EF58F2AF-1F2A-4221-BA74-A5C3076EFBDF' -resultBundlePath 'fastlane/test_output/OurScheme/OurScheme.xcresult' -enableCodeCoverage YES -testPlan 'Integration' GCC_TREAT_WARNINGS_AS_ERRORS=YES SWIFT_TREAT_WARNINGS_AS_ERRORS=YES OTHER_SWIFT_FLAGS='$(inherited) -Xfrontend -warn-long-function-bodies=700 -Xfrontend -warn-long-expression-type-checking=700' test | tee '/usr/local/var/buildkite-agent/builds/Proxy-1/us/pipeline/fastlane/output/App-OurScheme.log' | xcbeautify
```
</details>

There's only one result bundle generated which has the expected number of tests: 

<img width="97" alt="Screenshot 2022-01-14 at 19 54 18" src="https://user-images.githubusercontent.com/465802/149576859-5ec5fb03-3a47-4f10-a949-163ddaac9ce5.png">
 
Happy to provide any more info if that's helpful 🙂

joshdholtz commented:
> > > First up, thanks for adding this! We've been using a workaround to enable `xcbeautify` with `fastlane` (as we use parallel UI testing extensively) for a while.
> > > I've removed our workarounds and switched over to the `rc1` tag. I'm seeing this correctly detect success and failure states and show output correctly using `xcbeautify` as with our previous workaround. The only piece of odd behaviour I've seen so far is at the end of `scan` it prints:
> > > `Tests Passed: 0 failed, 9 total (0.669 seconds)`
> > > ... where in reality it's executed a few thousand tests and taken (sadly) a lot longer than that.
> > 
> > 
> > @sebskuse Weird! I wonder if its picking up a different `xcresult` somewhere 🤔 Are you able to send me all the options you are setting on `scan` and possibly your console output if you can?
> 
> This one was on one of our ephemeral CI machines, so no possibility of anything else being picked up. Can provide you with some more info though! Have attached the scan table below. Another thing I've noticed is because I've now removed output_types (previously we had this set to `output_types('')` as a workaround - presumed we no longer need this?) it's now being set to `junit,html`. Not sure if that matters for anything.
> 
> scan output
> which results in the following xcodebuild command:
> 
> xcodebuild
> There's only one result bundle generated which has the expected number of tests:
> 
> <img alt="Screenshot 2022-01-14 at 19 54 18" width="97" src="https://user-images.githubusercontent.com/465802/149576859-5ec5fb03-3a47-4f10-a949-163ddaac9ce5.png">
> 
> Happy to provide any more info if that's helpful 🙂

@sebskuse I just realized that `Tests Passed: 0 failed, 9 total (0.669 seconds)` is actually output provided by `xcodebuild` is actually output from [xbeautify](https://github.com/tuist/xcbeautify) 👉  https://github.com/tuist/xcbeautify/blob/b3fc2e05946397b056443df07bede3b0ab4e792c/Sources/XcbeautifyLib/TestSummary.swift#L19-L25 

So might be good to go report that issue there 😊 

On the other hand... what did `scan`'s result table show? Did that show the expected number?

<img width="868" alt="Screen Shot 2022-01-14 at 4 17 39 PM" src="https://user-images.githubusercontent.com/401294/149592551-6ad0ca6e-516d-4ae3-8280-0e839364976b.png">

joshdholtz committed:
Fix no formatter stuff

gobetti commented:
hi! thank you so much for this ❤️ I'll be able to test it better on Tuesday. One minor thing I can anticipate though is that the release changelog was my primary source of understanding the changes, and from that I understood xcbeautify would be used by default, I didn't realize I'd have to install it (for the same reason I haven't had to install xcpretty 😄). As obvious as it might be, just figured if it bit me it could bite others - might not hurt to mention that requirement. Thanks again!

joshdholtz commented:
> hi! thank you so much for this ❤️ I'll be able to test it better on Tuesday. One minor thing I can anticipate though is that the release changelog was my primary source of understanding the changes, and from that I understood xcbeautify would be used by default, I didn't realize I'd have to install it (for the same reason I haven't had to install xcpretty 😄). As obvious as it might be, just figured if it bit me it could bite others - might not hurt to mention that requirement. Thanks again!

@gobetti My pleasure! I'll add that requirement to the PR description above and in future release information but I'm also planning on creating a separate page on the docs for all xcpretty and xcbeautify information and settings 😊

gobetti commented:
I have one more quick report, I haven't figured what's going on but since installing xcbeautify, I'm not getting any output files anymore, with xcpretty I was getting html and junit (nearly empty, though). Now the default `test_output` directory doesn't get created at all. I'm not customizing any of the `*output*` options in `scan`, here's my configuration:
```
+------------------------------------------------+-----------------------------------------------------------------------------------------------+
|                                                          Summary for scan 2.201.0.rc1                                                          |
+------------------------------------------------+-----------------------------------------------------------------------------------------------+
| test_without_building                          | true                                                                                          |
| workspace                                      | redacted                                                                      |
| scheme                                         | redacted                                                                                  |
| testplan                                       | no_ui                                                                                         |
| derived_data_path                              | redacted/fastlane/builds_52a663cd609a34c52519f5735bd165c0a3c8fa77 |
| devices                                        | redacted                                           |
| prelaunch_simulator                            | true                                                                                          |
| xcargs                                         | redacted                                         |
| buildlog_path                                  | ~/Library/Logs/scan                                                                           |
| skip_slack                                     | true                                                                                          |
| skip_detect_devices                            | false                                                                                         |
| ensure_devices_found                           | false                                                                                         |
| force_quit_simulator                           | false                                                                                         |
| reset_simulator                                | false                                                                                         |
| disable_slide_to_type                          | true                                                                                          |
| reinstall_app                                  | false                                                                                         |
| clean                                          | false                                                                                         |
| open_report                                    | false                                                                                         |
| output_directory                               | ./fastlane/test_output                                                                        |
| output_types                                   | html,junit                                                                                    |
| include_simulator_logs                         | false                                                                                         |
| xcodebuild_formatter                           | xcbeautify                                                                                    |
| should_zip_build_products                      | false                                                                                         |
| output_xctestrun                               | false                                                                                         |
| result_bundle                                  | false                                                                                         |
| use_clang_report_name                          | false                                                                                         |
| disable_concurrent_testing                     | false                                                                                         |
| skip_build                                     | false                                                                                         |
| slack_use_webhook_configured_username_and_icon | false                                                                                         |
| slack_username                                 | fastlane                                                                                      |
| slack_icon_url                                 | https://fastlane.tools/assets/img/fastlane_icon.png                                           |
| slack_only_on_failure                          | false                                                                                         |
| xcodebuild_command                             | env NSUnbufferedIO=YES xcodebuild                                                             |
| skip_package_dependencies_resolution           | false                                                                                         |
| disable_package_automatic_updates              | false                                                                                         |
| use_system_scm                                 | false                                                                                         |
| number_of_retries                              | 0                                                                                             |
| fail_build                                     | true                                                                                          |
| xcode_path                                     | /Applications/Xcode.app  
```

joshdholtz commented:
@gobetti That will be fixed in `rc2` which I plan to release in about 10 minutes 😊

joshdholtz commented:
@chojnac @Kylmakalle I'm not able to recreate either of your crashes with `test_without_build: true`. Would either of you be able to give me what your `scan` configuration looks like and some console output? How many tests did you expect to be run?

joshdholtz committed:
Fix issues where output files were not being created

joshdholtz commented:
❇️  OFFICIAL UPDATE
Just released version `2.201.0.rc2` 👉 https://github.com/fastlane/fastlane/releases/tag/2.201.0.rc2
Major fix for `output_types` but did not fix the crash that was being experienced with `test_without_build: true` yet

gobetti commented on scan/lib/scan/runner.rb:
```diff
@@ -135,27 +137,118 @@ def retryable_tests(input)
+
+      formatter = Scan.config[:xcodebuild_formatter].chomp
+
+      if output_html? && formatter != 'xcpretty'
+        UI.error("HTML output is only available with `xcodebuild_formatter: 'xcpretty'` right now...")
```
nit: I'm getting this error even though I haven't explicitly chosen the html type, since it's in the default output_types. We can silence it by being explicit with just junit, but would it make sense to default that option according to the formatter?

gobetti replied:
actually, I'm also seeing this when I don't have xcbeautify installed and xcpretty was selected by default as the xcodebuild_formatter (but no explicit formatter nor output_types choices made by me)

joshdholtz replied:
Ahhhh, I'll remove this! I thought it would be helpful but seems like its probably more error prone that anything 🤷‍♂️

gobetti commented:
@joshdholtz you've answered so fast that I had to try rc2 today 😅 so the main blocker for me I'm seeing is that the report.junit file that's in the output_directory is mostly empty (and yep I'm using xcbeautify, see my previously attached scan config, also confirmed from the xcodebuild command printed by fastlane)

output report.unit:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="0" failures="0">
</testsuites>
```
before this release we've been using disable_xcpretty=true and trainer manually installed as a plugin and the way we use it, it outputs an XML file which has been working fine to `junit.parse` and read `junit.failures` for example

chojnac commented:
> @chojnac @Kylmakalle I'll get a fix out for `test_without_build: true` into `rc2` for you 💪 I'll ping you when its released!

@joshdholtz  I'm happy to report that the latest release(`rc2`) resolved all issues reported by me. Good job, and thanks!

joshdholtz commented:
> @joshdholtz you've answered so fast that I had to try rc2 today 😅 so the main blocker for me I'm seeing is that the report.junit file that's in the output_directory is mostly empty (and yep I'm using xcbeautify, see my previously attached scan config, also confirmed from the xcodebuild command printed by fastlane)
> 
> output report.unit:
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <testsuites tests="0" failures="0">
> </testsuites>
> ```
> 
> before this release we've been using disable_xcpretty=true and trainer manually installed as a plugin and the way we use it, it outputs an XML file which has been working fine to `junit.parse` and read `junit.failures` for example
> 
> edit: quick thought, could it be that the junit file being output is still the old one, not generated by trainer at all?

@gobetti Hmm, that is really weird you are seeing zero tests! 

1) What number of tests and failures are you seeing in `scan`'s "Test Results" table?
2) Can you send me the scan `scan` configuration that you're using?

gobetti commented:
**1.** first, I use `build_for_testing: true` i.e. not running tests initially, at this point the table prints `Number of tests    | 0 |`, that's not really an issue and isn't related to this PR, just mentioning for completeness. Then I run `test_without_building: true` which behaves differently according to having test failures or not:
**a.** when all tests are passing: the table numbers match the xcresult's (times 2 given I run to 2 devices). The report.junit matches that as well.
```
[00:26:38]: ▸ Test Execute Succeeded
[00:26:39]: HTML output is only available with `xcodebuild_formatter: 'xcpretty'` right now...
+--------------------+------+
|       Test Results        |
+--------------------+------+
| Number of tests    | 2676 |
| Number of failures | 0    |
+--------------------+------+

[00:26:39]: Cruising back to lane 'ios test' 🚘
```
output report.junit:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="2676" failures="0">
...
```
**b.** after adding XCTFail() to a couple tests, the `test_without_building` run finishes without printing a table at all, and a nearly empty report.junit file is output. These are the final lines printed by the `run_steps` step before it goes to the next one (in my case, `danger`):
```
** TEST EXECUTE FAILED **
[00:19:42]: Exit status: 65
[00:19:42]: --------------------
[00:19:42]: --- Step: danger ---
[00:19:42]: --------------------
```
**c.** for debugging purposes I put my 2 separate calls back into 1 by not using build_for_testing/test_without_building, in this case no "Test Results" tables and no report.junit file are generated. The conclusion here is that the nearly empty report.junit file I'm seeing in `b` was generated by the initial `build_for_testing ` run (remember the first 0 tests table?), and just wasn't overwritten by `test_without_building`

----

My guess from those investigations is that we need to `catch` errors from `scan` and run trainer regardless. That's what I do currently with my manual call to trainer. Something like (one liner for short) `begin` run_tests `ensure` trainer `end`. Am I the only one seeing this behavior when there are test failures?

----

**2.** [my scan summary is here](https://github.com/fastlane/fastlane/pull/19629#issuecomment-1013584230) the only caveat to that is that I run that scan configuration twice, first with `build_for_testing: true`, second with `test_without_building: true`. There might be copyright infringement if I share more than that, but I'm hopeful point `1` has given good insights, also I'll take a stab at fixing this myself 🤞 given I have the environment to repro the issue and a couple first guesses. Thanks again!

gobetti commented on scan/lib/scan/runner.rb:
```diff
@@ -135,27 +137,118 @@ def retryable_tests(input)
+      return results
+    end
+
+    def handle_results(tests_exit_status)
+      results = trainer_test_results
```
this isn't reached for me when any tests fail, because `test_app` in this file is raising an exception instead of returning an exit status, and this existing code doesn't handle exception:
```
    def run
      handle_results(test_app)
```
if it did then I do get a fulfilled test results table and report.junit file, for example:
```
    def run
      begin
        results = test_app
      rescue => ex
      ensure
        handle_results(results || -1)
        raise ex unless ex.nil?
```
but I guess we'd want to fix `test_app` instead

gobetti replied:
note though that not raising an exception anymore may be a breaking change for some... I myself wrap `run_tests` with `begin` (so far based on empirical observations) so my `trainer` call runs for failed builds too (well, that's when we most want it, right 😄)

joshdholtz replied:
@gobetti Ohhhhh, this makes so much sense now!! Can you try setting `fail_build: false` and see if that "fixes" it for you? 😇

I totally see how this existing fail logic doesn't make sense anymore. I will look into! Thanks so much for all the detail 🙌

gobetti replied:
ha! glad you asked, because I considered making that a **d.** case in my huge post - no, adding `fail_build: false` doesn't make a difference. But given I can't find `fail_build` referenced at all in scan/runner.rb I didn't know that was worth consideration

gobetti replied:
I see now where `fail_build` is used in fastlane's code but it's _outside_ to the `run` method I shared above, so it makes sense that test_app still raises and consequently skips handle_results/trainer_test_results

also fyi, not caused by this PR but affects the desired behavior here - although my failure was just tests failing (exit status 65) the exception is coming as a FastlaneBuildFailure in which case `fail_build` is ignored in the current code:
```
        rescue FastlaneCore::Interface::FastlaneBuildFailure => ex
          # Specifically catching FastlaneBuildFailure to prevent build/compile errors from being
          # silenced when :fail_build is set to false
          # :fail_build should only suppress testing failures
          raise ex
```

gobetti replied:
the build failure is defined in scan/error_handler.rb:
```
        when /Testing failed/
          UI.build_failure!("Error building the application. #{details}")
```
if I'm interpreting correctly, it sounds inconsistent to the code block I pasted on my previous message - `:fail_build should only suppress testing failures` seems to suggest a testing failure shouldn't be a FastlaneBuildFailure

gobetti replied:
okkk I think I've narrowed it down 🙌 the logic in `scan/error_handler.rb` was apparently tailored for xcpretty and still works well if I force `xcodebuild_formatter: 'xcpretty'` - however if I use xcbeautify or `xcodebuild_formatter: ''`, then the `output` that is used by `handle_build_error` is something that doesn't work with the logic currently in place

gobetti replied:
https://github.com/fastlane/fastlane/issues/19827

joshdholtz commented:
@sebskuse I made a PR into xcbeautify that fixes that summary count that you were seeing that was wrong 👉 https://github.com/tuist/xcbeautify/pull/65

Kylmakalle commented:
Issues with `test_without_building: true` fixed for me in rc2. Thanks.

AvdLee reviewed (approved):
Haven't been able to test this locally myself, but approving the code already. I've seen many other testing the implementation, which is enough for me for now. I'm hoping to try this out soon on our implementation as well. Thanks!

chojnac commented:
This may be off-topic in the context of this PR, but because `trainer` is now part of the scan command, I want to mention this issue.
 In the case of "test repetition" `trainer` incorrectly generates failures count in xml file. In the final summary for a test suite, `trainer` counts failure even if later the test was successfully completed.
This flaw makes CI fails. So for those of us running unit tests on CI this is a big issue for fully enabling test repetition.
An example build log:
```
▸ Test Suite ExampleTests started on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)'
▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (38.450 seconds)
▸     ✔ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (32.767 seconds)
```
what I get:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="91" failures="1">
    <testsuite name="UITests" tests="1" failures="1" time="2086.6292536258698">
        <testcase classname="ExampleTests" name="test01()" time="38.450246930122375">
            <failure message="failed - Timeout some elements missing (UITestFile.swift#CharacterRangeLen=0&amp;EndingLineNumber=280)">
            </failure>
        </testcase>
        <testcase classname="ExampleTests" name="test01()" time="32.76705300807953">
        </testcase>
    </testsuite>
</testsuites>
```
what I expect
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="91" failures="1">
    <testsuite name="UITests" tests="1" failures="0" time="2086.6292536258698">
        <testcase classname="ExampleTests" name="test01()" time="38.450246930122375">
            <failure message="failed - Timeout some elements missing (UITestFile.swift#CharacterRangeLen=0&amp;EndingLineNumber=280)">
            </failure>
        </testcase>
        <testcase classname="ExampleTests" name="test01()" time="32.76705300807953">
        </testcase>
    </testsuite>
</testsuites>
```

chojnac commented:
> @chojnac This is totally on topic! Are you using the `number_of_retries` option to retry on failed tests?

This is my `scan` command:

```
scan(
    test_without_building: true,
    scheme: "ApplicationUITests",
    reset_simulator: true,
    disable_slide_to_type: true,
    devices: ["iPhone 11 Pro"],
    testplan: "AllUITests",
    max_concurrent_simulators: 2,
    code_coverage: false,
    xcodebuild_formatter: "xcbeautify",
    output_types: "junit",
    number_of_retries: 3,
    result_bundle: true
)
```
also the "Execute in parallel" flag is enabled in the test plan.

joshdholtz committed:
Don't run trainer when build_for_testing is set and show output types as warnings with how to get rid of them

joshdholtz committed:
Fix for trainer plist parser

joshdholtz commented:
@chojnac Thank you so much for the sample project! I think `2.201.0.rc3` should have all the fixes you need 🤞

osrufung commented:
Not sure why... but trainer seems is failing in my case because `config[:extension]` is not defined. I'll try to provide a example project asap

```

[2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
[2022-01-18T07:39:26.196Z] |                       Summary for trainer 0.9.1                        |
[2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
[2022-01-18T07:39:26.196Z] | path             | ./fastlane/test_output/MyApp.xcresult                |
[2022-01-18T07:39:26.196Z] | silent           | true                                                |
[2022-01-18T07:39:26.196Z] | output_filename  | report.junit                                        |
[2022-01-18T07:39:26.196Z] | output_directory | /Users/MyUser/workspace/repository/fastlane/test_output |
[2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
[2022-01-18T07:39:26.196Z] 
[2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
[2022-01-18T07:39:26.196Z] |                  Lane Context                   |
[2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
[2022-01-18T07:39:26.196Z] | DEFAULT_PLATFORM           | ios                |
[2022-01-18T07:39:26.196Z] | PLATFORM_NAME              | ios                |
[2022-01-18T07:39:26.196Z] | LANE_NAME                  | ios build_and_test |
[2022-01-18T07:39:26.196Z] | SCAN_DERIVED_DATA_PATH     | derived_data       |
[2022-01-18T07:39:26.196Z] | SCAN_GENERATED_PLIST_FILES | []                 |
[2022-01-18T07:39:26.196Z] | SCAN_GENERATED_PLIST_FILE  |                    |
[2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
[2022-01-18T07:39:26.196Z] [08:39:35]: no implicit conversion of nil into String

```

osrufung commented:
> Not sure why (that's happening in rc2 and rc3) ... but trainer seems is failing in my case because `config[:extension]` is not defined. I'll try to provide a example project asap and try to fix it by providing this as environment variable.
> 
> ```
> 
> [2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
> [2022-01-18T07:39:26.196Z] |                       Summary for trainer 0.9.1                        |
> [2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
> [2022-01-18T07:39:26.196Z] | path             | ./fastlane/test_output/MyApp.xcresult                |
> [2022-01-18T07:39:26.196Z] | silent           | true                                                |
> [2022-01-18T07:39:26.196Z] | output_filename  | report.junit                                        |
> [2022-01-18T07:39:26.196Z] | output_directory | /Users/MyUser/workspace/repository/fastlane/test_output |
> [2022-01-18T07:39:26.196Z] +------------------+-----------------------------------------------------+
> [2022-01-18T07:39:26.196Z] 
> [2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
> [2022-01-18T07:39:26.196Z] |                  Lane Context                   |
> [2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
> [2022-01-18T07:39:26.196Z] | DEFAULT_PLATFORM           | ios                |
> [2022-01-18T07:39:26.196Z] | PLATFORM_NAME              | ios                |
> [2022-01-18T07:39:26.196Z] | LANE_NAME                  | ios build_and_test |
> [2022-01-18T07:39:26.196Z] | SCAN_DERIVED_DATA_PATH     | derived_data       |
> [2022-01-18T07:39:26.196Z] | SCAN_GENERATED_PLIST_FILES | []                 |
> [2022-01-18T07:39:26.196Z] | SCAN_GENERATED_PLIST_FILE  |                    |
> [2022-01-18T07:39:26.196Z] +----------------------------+--------------------+
> [2022-01-18T07:39:26.196Z] [08:39:35]: no implicit conversion of nil into String
> ...
> /Users/MyUser/.rvm/gems/ruby-3.1.0/gems/trainer-0.9.1/lib/trainer/test_parser.rb:41:in `gsub': \e[31m[!] no implicit conversion of nil into String\e[0m (TypeError)
> 
> from /Users/MyUser/.rvm/gems/ruby-3.1.0/gems/trainer-0.9.1/lib/trainer/test_parser.rb:41:in `block in auto_convert'
> ```

ok, found the **conflict**. I was already including trainer in my `Gemfile`. Removed and using the included version with fastlane.

joshdholtz commented:
@osrufung Ah yes, sorry about that! I should have included that in my instructions 🤦‍♂️ I need to find a way to see if I can add an error if it's being included from the external gem 🙃 But I will include removing that in the migration instructions!

osrufung commented:
> @osrufung Ah yes, sorry about that! I should have included that in my instructions 🤦‍♂️ I need to find a way to see if I can add an error if it's being included from the external gem 🙃 But I will include removing that in the migration instructions!

Don't worry @joshdholtz that's something easy to figure out. What I can confirm is that after dealing with some little issues with .xml/.junit extension, we are able to process the report with unit in Jenkins and see red/orange/green jobs. 🎉

Thank you for this RC 👏

chojnac commented:
Hey, I have one more case related to the test repetition that I would like to discuss. 
When I run my flaky UI test suite, rerunning failing tests solves issues with flakiness and allows "lane" to succeed. 
However, when CI (TeamCity) process the junit report,  any failures in XML file result in the whole job failing. CI report generator is not very sophisticated, and when it sees failures in the file, it just fails the entire job. 

I'm not sure if this is specific to my case or a more generic problem with  CI systems. Probably I could be solved by post-processing junit file, but I think it might be helpful for others to address this case in 'trainer'.
So I want to suggest an option to `trainer` / `scan` to report only failed (in all repeating runs) tests cases. From what I see, there is already some code in `trainer` that might be used to build this feature. 

To visualise how this option might work:
```
▸ Test Suite ExampleTests started on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)'
▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (38.450 seconds)
▸     ✔ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (32.767 seconds)
```
will generate
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="1" failures="0">
    <testsuite name="UITests" tests="1" failures="0" time="2086.6292536258698">
        <testcase classname="ExampleTests" name="test01()" time="32.76705300807953">
        </testcase>
    </testsuite>
</testsuites>
```

and 
```
▸ Test Suite ExampleTests started on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)'
▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (38.450 seconds)
▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (32.767 seconds)
```
will produce
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="1" failures="1">
    <testsuite name="UITests" tests="1" failures="1" time="2086.6292536258698">
        <testcase classname="ExampleTests" name="test01()" time="38.450246930122375">
            <failure message="failed - Timeout some elements missing (UITestFile.swift#CharacterRangeLen=0&amp;EndingLineNumber=280)">
            </failure>
        </testcase>
        </testcase>
    </testsuite>
</testsuites>
```

sebskuse commented:
> Hey, I have one more case related to the test repetition that I would like to discuss. 
> 
> When I run my flaky UI test suite, rerunning failing tests solves issues with flakiness and allows "lane" to succeed. 
> 
> However, when CI (TeamCity) process the junit report,  any failures in XML file result in the whole job failing. CI report generator is not very sophisticated, and when it sees failures in the file, it just fails the entire job. 
> 
> 
> 
> I'm not sure if this is specific to my case or a more generic problem with  CI systems. Probably I could be solved by post-processing junit file, but I think it might be helpful for others to address this case in 'trainer'.
> 
> So I want to suggest an option to `trainer` / `scan` to report only failed (in all repeating runs) tests cases. From what I see, there is already some code in `trainer` that might be used to build this feature. 
> 
> 
> 
> To visualise how this option might work:
> 
> ```
> 
> ▸ Test Suite ExampleTests started on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)'
> 
> ▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (38.450 seconds)
> 
> ▸     ✔ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (32.767 seconds)
> 
> ```
> 
> will generate
> 
> ```xml
> 
> <?xml version="1.0" encoding="UTF-8"?>
> 
> <testsuites tests="1" failures="0">
> 
>     <testsuite name="UITests" tests="1" failures="0" time="2086.6292536258698">
> 
>         <testcase classname="ExampleTests" name="test01()" time="32.76705300807953">
> 
>         </testcase>
> 
>     </testsuite>
> 
> </testsuites>
> 
> ```
> 
> 
> 
> and 
> 
> ```
> 
> ▸ Test Suite ExampleTests started on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)'
> 
> ▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (38.450 seconds)
> 
> ▸     ✖ test01 on 'Clone 3 of iPhone 11 Pro - UITests-Runner (66463)' (32.767 seconds)
> 
> ```
> 
> will produce
> 
> ```xml
> 
> <?xml version="1.0" encoding="UTF-8"?>
> 
> <testsuites tests="1" failures="1">
> 
>     <testsuite name="UITests" tests="1" failures="1" time="2086.6292536258698">
> 
>         <testcase classname="ExampleTests" name="test01()" time="38.450246930122375">
> 
>             <failure message="failed - Timeout some elements missing (UITestFile.swift#CharacterRangeLen=0&amp;EndingLineNumber=280)">
> 
>             </failure>
> 
>         </testcase>
> 
>         </testcase>
> 
>     </testsuite>
> 
> </testsuites>
> 
> ```
> 
> 
> 
> 

To add another datapoint to this: We had the same issue with danger-junit (which we use to report on success / failure on PR builds). 

I ended up forking in the end to add functionality that identifies flaky tests over failures and reports them into PRs rather than failing the build - keeping the JUnit intact so you could see the underlying reasons for the flakes.

joshdholtz committed:
Give error about removing trainer plugin

joshdholtz committed:
Add new option to scan to ignore retry attempts in output results

joshdholtz committed:
Improve trainer remove retry attempt count

joshdholtz committed:
Added link to doc

joshdholtz committed:
Change version back to 2.200.0
